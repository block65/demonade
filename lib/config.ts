import { CustomError, Status } from '@block65/custom-error';
import { findUp } from 'find-up';
import { lilconfig } from 'lilconfig';
import { dirname, relative } from 'node:path';
import { logger } from './logger.js';

/**
 * @private
 */
export interface InternalConfig {
  command: string;
  args: string[];
  watch: string[];
  ignore?: (string | RegExp | ((path: string) => boolean))[];
  workingDirectory: string;
  signal: NodeJS.Signals;
  delay: number;
  quiet?: boolean;
  verbose?: boolean;
}

export interface CliConfig {
  command?: string;
  args?: string[];
  signal?: NodeJS.Signals;
  watch?: string[];
  ignore?: string[];
  quiet?: boolean;
  verbose?: boolean;
  delay?: number;
}

export interface Config {
  command?: string;
  args?: string[];
  watch?: string[];
  ignore?: (string | RegExp | ((path: string) => boolean))[];
  signal?: NodeJS.Signals;
  quiet?: boolean;
  verbose?: boolean;
  delay?: number;
}

class ConfigError extends CustomError {
  public code = Status.INVALID_ARGUMENT;
}

function loadEsm(specifier: string) {
  return import(specifier);
}

export async function resolveConfig(
  cliConfig: CliConfig,
): Promise<InternalConfig> {
  const configResult = await lilconfig('demonade', {
    loaders: {
      '.js': loadEsm,
      '.mjs': loadEsm,
    },
  }).search();

  if (
    !cliConfig.command &&
    !cliConfig.args &&
    !configResult?.config.command &&
    !configResult?.config.args
  ) {
    throw new ConfigError('Provide either `command` or `args`').debug({
      config: configResult?.config,
      cliArgs: cliConfig,
    });
  }

  const packageJson = await findUp('package.json', {
    type: 'file',
    allowSymlinks: false,
  });
  const packageDir = dirname(packageJson || process.cwd());

  const workingDirectory = configResult?.filepath
    ? dirname(configResult?.filepath)
    : packageDir;

  const watch = [
    ...new Set<string>([
      ...(configResult?.config?.watch || []),
      ...(cliConfig?.watch || []),
      ...(!configResult?.config?.watch && cliConfig?.watch
        ? [workingDirectory]
        : []),
    ]),
  ]
    .map((p) => relative(workingDirectory, p))
    .filter(Boolean);

  const configIgnore: (string | RegExp | ((path: string) => boolean))[] = [
    ...new Set([
      ...(configResult?.config?.ignore || []),
      ...(cliConfig?.ignore || []),
    ]),
  ].filter(Boolean);

  const defaultIgnore = [/(^|[/\\])\../, '**/node_modules/**'];

  const ignore = configIgnore.length > 0 ? configIgnore : defaultIgnore;

  const resolved: InternalConfig = {
    workingDirectory,
    watch,
    ignore,
    command:
      cliConfig?.command || configResult?.config?.command || process.execPath,
    args: cliConfig?.args || configResult?.config?.args || [],
    signal: cliConfig?.signal || configResult?.config?.signal || 'SIGUSR2',
    delay: cliConfig?.delay || configResult?.config?.delay || 200,
    quiet: cliConfig?.quiet || configResult?.config?.quiet,
    verbose: cliConfig?.verbose || configResult?.config?.verbose,
  };

  logger.trace(
    {
      cliConfig,
      resolved,
    },
    'resolved config',
  );

  return resolved;
}
