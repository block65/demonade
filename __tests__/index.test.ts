import { describe, expect, test } from '@jest/globals';
import { resolveConfig } from '../lib/config.js';

describe('configs', () => {
  test('Full fat', async () => {
    await expect(
      resolveConfig({
        args: ['-c', 'hello'],
        command: 'woot',
        exclude: ['node_modules'],
        include: ['dist/**/*.js'],
        delay: 12,
        signal: 'SIGABRT',
        verbose: true,
      }),
    ).resolves.toMatchSnapshot<any>({
      workingDirectory: expect.any(String),
    });
  });
});
