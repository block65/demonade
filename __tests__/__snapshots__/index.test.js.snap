// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configs Full fat 1`] = `
{
  "args": [
    "-c",
    "hello",
  ],
  "command": "woot",
  "delay": 12,
  "ignore": [
    "node_modules",
  ],
  "quiet": undefined,
  "signal": "SIGABRT",
  "verbose": undefined,
  "watch": [
    "dist/**/*.js",
  ],
  "workingDirectory": Any<String>,
}
`;
